workspaces/diff-engine/src/events/spec_chunk.rs

```diff
@@ -0,0 +1,332 @@
+  }
+}
+
+impl Event for BatchChunkEvent {
+  fn event_type(&self) -> &'static str {
+    "BatchChunkEvent"
+  }
+}
+
+impl Event for UnknownChunkEvent {
+  fn event_type(&self) -> &'static str {
+    "UnknownChunkEvent"
+  }
+}
+
+// TODO: Implement this for an impl Iterator<Item=SpecEvent> rather than requiring a Vec
+impl From<(String, bool, Vec<SpecEvent>)> for SpecChunkEvent {
+  fn from((name, is_root, events): (String, bool, Vec<SpecEvent>)) -> Self {
+    if is_root {
+      Self::Root(RootChunkEvent {
+        id: String::from("root"),
+        name,
+        events,
+      })
+    } else {
+      match BatchChunkEvent::try_from((name, events)) {
+        Ok(batch_chunk) => Self::Batch(batch_chunk),
+        Err((msg, name, events)) => Self::Unknown(UnknownChunkEvent { name, events }),
+      }
+    }
+  }
+}
+
+// TODO: Implement this for an impl Iterator<Item=SpecEvent> rather than requiring a Vec
+impl TryFrom<(String, Vec<SpecEvent>)> for BatchChunkEvent {
+  // Give ownership of name and events back when we fail
+  type Error = (&'static str, String, Vec<SpecEvent>); // TODO: replace with proper error type
+
+  fn try_from((name, events): (String, Vec<SpecEvent>)) -> Result<Self, Self::Error> {
+    let batch_ids = parse_batch_chunk_events(&events);
+
+    match batch_ids {
+      Ok((id, parent_id)) => Ok(Self {
+        id,
+        name,
+        parent_id,
+        events,
+      }),
+      Err(err) => Err((err, name, events)),
+    }
+  }
+}
+
+fn parse_batch_chunk_events(events: &Vec<SpecEvent>) -> Result<(String, String), &'static str> {
+  let first_event = events.first().ok_or("Chunk does not have any events")?;
+
+  let batch_start_event = match first_event {
+    SpecEvent::RfcEvent(RfcEvent::BatchCommitStarted(e)) => Ok(e),
+    _ => Err("Chunk does not start with a BatchCommitStarted event"),
+  }?;
+
+  let batch_id = batch_start_event.batch_id.clone();
+  let parent_id = match &batch_start_event.parent_id {
+    Some(parent_id) => Ok(parent_id.clone()),
+    _ => Err("BatchCommitStarted does not have a parent_id"),
+  }?;
+
+  let last_event = events.last().ok_or("Chunk does not have any events")?;
+
+  let batch_end_event = match last_event {
+    SpecEvent::RfcEvent(RfcEvent::BatchCommitEnded(e)) => Ok(e),
+    _ => Err("Chunk does not end in a BatchCommitEnded event"),
+  }?;
+
+  if batch_end_event.batch_id != batch_id {
+    Err("BatchCommitEnded event does not have matching parent_id")
+  } else {
+    Ok(())
+  }?;
+
+  let intermediate_events = events.iter().skip(1).take(events.len() - 2); // skip the first, don't take the last
+  for event in intermediate_events {
+    if let SpecEvent::RfcEvent(RfcEvent::BatchCommitStarted(e)) = event {
+      return Err("Chunk cannot include nested BatchCommitStarted events");
+    } else if let SpecEvent::RfcEvent(RfcEvent::BatchCommitEnded(e)) = event {
+      return Err("Chunk cannot include nested BatchCommitEnded events");
+    }
+  }
+
+  Ok((batch_id, parent_id))
+}
```

Comment:
This is where the logic that determines what a valid `Batch` chunk is. Made sure we have descriptive errors here, so we can debug it easily.